src/server/lib/callback-url-handler.js

```diff
@@ -1,32 +1,42 @@
-import * as cookie from '../lib/cookie'
+// @ts-check
+import * as cookie from "../lib/cookie"
 
 /**
  * Get callback URL based on query param / cookie + validation,
  * and add it to `req.options.callbackUrl`.
- * @note: `req.options` must already be defined when called.
+ * @type {import("types/internals").NextAuthApiHandler}
  */
-export default async function callbackUrlHandler (req, res) {
+export default async function callbackUrlHandler(req, res) {
   const { query } = req
   const { body } = req
-  const { cookies, baseUrl, defaultCallbackUrl, callbacks } = req.options
+  const { cookies, baseUrl, callbacks } = req.options
 
-  // Handle preserving and validating callback URLs
-  // If no defaultCallbackUrl option specified, default to the homepage for the site
-  let callbackUrl = defaultCallbackUrl || baseUrl
+  let callbackUrl = baseUrl
   // Try reading callbackUrlParamValue from request body (form submission) then from query param (get request)
   const callbackUrlParamValue = body.callbackUrl || query.callbackUrl || null
   const callbackUrlCookieValue = req.cookies[cookies.callbackUrl.name] || null
   if (callbackUrlParamValue) {
     // If callbackUrl form field or query parameter is passed try to use it if allowed
-    callbackUrl = await callbacks.redirect(callbackUrlParamValue, baseUrl)
+    callbackUrl = await callbacks.redirect({
+      url: callbackUrlParamValue,
+      baseUrl,
+    })
   } else if (callbackUrlCookieValue) {
     // If no callbackUrl specified, try using the value from the cookie if allowed
-    callbackUrl = await callbacks.redirect(callbackUrlCookieValue, baseUrl)
+    callbackUrl = await callbacks.redirect({
+      url: callbackUrlCookieValue,
+      baseUrl,
+    })
   }
 
-  // Save callback URL in a cookie so that can be used for subsequent requests in signin/signout/callback flow
-  if (callbackUrl && (callbackUrl !== callbackUrlCookieValue)) {
-    cookie.set(res, cookies.callbackUrl.name, callbackUrl, cookies.callbackUrl.options)
+  // Save callback URL in a cookie so that it can be used for subsequent requests in signin/signout/callback flow
+  if (callbackUrl && callbackUrl !== callbackUrlCookieValue) {
```

Comment:
I have the impression we don't need the first boolean check on `callbackUrl`, as it's [going to be always truthy](https://github.com/nextauthjs/next-auth/blob/main/src/lib/parse-url.js#L23).

Reply:
This is internal code, I would do it in a different PR. I think the whole `callbackUrlHandler` could be refactored. I'll have more confidence when we have tests at place.